x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/ml/datafeed/DatafeedConfig.java

```diff
@@ -0,0 +1,344 @@
+
+        public Builder(String id, String jobId) {
+            this.id = Objects.requireNonNull(id, ID.getPreferredName());
+            this.jobId = Objects.requireNonNull(jobId, Datafeed.JOB_ID.getPreferredName());
+        }
+
+        public Builder(DatafeedConfig config) {
+            this.id = config.id;
+            this.jobId = config.jobId;
+            this.queryDelay = config.queryDelay;
+            this.frequency = config.frequency;
+            this.indices = config.indices;
+            this.types = config.types;
+            this.query = config.query;
+            this.aggregations = config.aggregations;
+            this.scriptFields = config.scriptFields;
+            this.scrollSize = config.scrollSize;
+            this.chunkingConfig = config.chunkingConfig;
+        }
+
+        public Builder setId(String datafeedId) {
+            id = Objects.requireNonNull(datafeedId, ID.getPreferredName());
+            return this;
+        }
+
+        public Builder setJobId(String jobId) {
+            this.jobId = Objects.requireNonNull(jobId, Datafeed.JOB_ID.getPreferredName());
+            return this;
+        }
+
+        public Builder setIndices(List<String> indices) {
+            this.indices = Objects.requireNonNull(indices, INDICES.getPreferredName());
+            return this;
+        }
+
+        public Builder setTypes(List<String> types) {
+            this.types = Objects.requireNonNull(types, TYPES.getPreferredName());
+            return this;
+        }
+
+        public Builder setQueryDelay(TimeValue queryDelay) {
+            this.queryDelay = queryDelay;
+            return this;
+        }
+
+        public Builder setFrequency(TimeValue frequency) {
+            this.frequency = frequency;
+            return this;
+        }
+
+        public Builder setQuery(QueryBuilder query) {
+            this.query = Objects.requireNonNull(query, QUERY.getPreferredName());
+            return this;
+        }
+
+        public Builder setAggregations(AggregatorFactories.Builder aggregations) {
+            this.aggregations = aggregations;
+            return this;
+        }
+
+        public Builder setScriptFields(List<SearchSourceBuilder.ScriptField> scriptFields) {
+            List<SearchSourceBuilder.ScriptField> sorted = new ArrayList<>(scriptFields);
+            sorted.sort(Comparator.comparing(SearchSourceBuilder.ScriptField::fieldName));
+            this.scriptFields = sorted;
+            return this;
+        }
+
+        public Builder setScrollSize(int scrollSize) {
+            this.scrollSize = scrollSize;
+            return this;
+        }
+
+        public Builder setChunkingConfig(ChunkingConfig chunkingConfig) {
+            this.chunkingConfig = chunkingConfig;
+            return this;
+        }
+
+        public DatafeedConfig build() {
+            Objects.requireNonNull(id, ID.getPreferredName());
+            Objects.requireNonNull(jobId, Datafeed.JOB_ID.getPreferredName());
```

Comment:
I don't think you need these `requireNonNull`s here, as every place where these fields could be changed already has them.

Reply:
```
DatafeedConfig.Builder builder = DatafeedConfig.Builder.new();
builder.build();
```
Is still possible. So, the two `String` fields would be `null` on build.

Reply:
I have opted to remove the empty constructor to enforce setting the job_id and the id.